3/assets/3.6-finished-code.rs

```diff
@@ -0,0 +1,315 @@
+
+            Self::_transfer_from(sender, to, kitty_id)?;
+
+            Ok(())
+        }
+
+        fn buy_kitty(origin, kitty_id: T::Hash, max_price: T::Balance) -> Result {
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+            let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+            ensure!(owner != sender, "You can't buy your own cat");
+
+            let mut kitty = Self::kitty(kitty_id);
+
+            let kitty_price = kitty.price;
+            ensure!(!kitty_price.is_zero(), "The cat you want to buy is not for sale");
+            ensure!(kitty_price <= max_price, "The cat you want to buy costs more than your max price");
+
+            <balances::Module<T>>::make_transfer(&sender, &owner, kitty_price)?;
+
+            Self::_transfer_from(owner.clone(), sender.clone(), kitty_id)?;
+
+            kitty.price = <T::Balance as As<u64>>::sa(0);
+            <Kitties<T>>::insert(kitty_id, kitty);
+
+            Self::deposit_event(RawEvent::Bought(sender, owner, kitty_id, kitty_price));
+
+            Ok(())
+        }
+
+        fn breed_kitty(origin, kitty_id_1: T::Hash, kitty_id_2: T::Hash) -> Result{
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id_1), "This cat 1 does not exist");
+            ensure!(<Kitties<T>>::exists(kitty_id_2), "This cat 2 does not exist");
+
+            let nonce = <Nonce<T>>::get();
+            let random_hash = (<system::Module<T>>::random_seed(), &sender, nonce)
+                .using_encoded(<T as system::Trait>::Hashing::hash);
+
+            let kitty_1 = Self::kitty(kitty_id_1);
+            let kitty_2 = Self::kitty(kitty_id_2);
+
+            let mut final_dna = kitty_1.dna;
+
+            for (i, (dna_2_element, r)) in kitty_2.dna.as_ref().iter().zip(random_hash.as_ref().iter()).enumerate() {
+                if r % 2 == 0 {
+                    final_dna.as_mut()[i] = *dna_2_element;
+                }
+            }
+
+            let new_kitty = Kitty {
+                id: random_hash,
+                dna: final_dna,
+                price: <T::Balance as As<u64>>::sa(0),
+                gen: cmp::max(kitty_1.gen, kitty_2.gen) + 1,
+            };
+
+            Self::_mint(sender, random_hash, new_kitty)?;
+
+            <Nonce<T>>::mutate(|n| *n += 1);
+
+            Ok(())
+        }
+
+        fn create_auction(origin, kitty_id: T::Hash, min_bid: T::Balance, expiry_in_minutes: u64) -> Result {
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+            let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+            ensure!(owner == sender, "You can't set an auction for a cat you don't own");
+
+            let expiry = <system::Module<T>>::block_number() + T::BlockNumber::sa(expiry_in_minutes*12);
```

Comment:
If you did want to use time directly, it would be an estimate, but you should pull the "block creation time" from the blockchain configuration or something, rather than having a hardcoded value. Someone else may add your module to their blockchain with a different block time

Reply:
I have not looked yet how to get that value into a module, but my guess is that it should be available in `system` if not, we should add it

Reply:
Block creation time is not needed anymore as I switch the expiry to be a `BlockNumber`. It makes sense since other modules can use this module, and the time to blocknumber conversion can be handled in the UI layer. 

In any case, I'd be curious to learn how we can get current block time though.

Reply:
The `timestamp` module has a storage item which you can access which has this data:

```
decl_storage! {
	trait Store for Module<T: Trait> as Timestamp {
		/// Current time for the current block.
		pub Now get(now) build(|_| T::Moment::sa(0)): T::Moment;
		/// The minimum (and advised) period between blocks.
		pub BlockPeriod get(block_period) config(period): T::Moment = T::Moment::sa(5);

		/// Did the timestamp get updated in this block?
		DidUpdate: bool;
	}
}
```

The `BlockPeriod`

Reply:
Thanks Shawn, I was able to read it as u64:

```rust
let block_period = <timestamp::Module<T>>::block_period().as_();
```